src/runtime/runtime.cs

```diff
@@ -522,11 +526,21 @@ private static void MoveClrInstancesOnwershipToPython()
                         PyObject_GC_Track(obj.pyHandle);
                     }
                 }
-                if (obj.gcHandle.IsAllocated)
+            }
+            foreach (var entry in copyObjs)
+            {
+                ManagedType obj = entry.Key;
+                if (!objs.ContainsKey(obj))
                 {
-                    obj.gcHandle.Free();
+                    System.Diagnostics.Debug.Assert(obj.gcHandle == default);
+                    continue;
+                }
+                if (obj.RefCount > 1)
+                {
+                    obj.FreeGCHandle();
+                    Marshal.WriteIntPtr(obj.pyHandle, ObjectOffset.magic(obj.tpHandle), IntPtr.Zero);
                 }
```

Comment:
Why can't you do this unconditionally? This condition makes reasoning about the state much harder.

Reply:
If `obj.RefCount == 1`, the next `XDecref` would dealloc the obj, thus it's no need to call these things.

Reply:
Is this similar to `tp_clear` in any way? I would suggest to extract these two updates (`FreeGCHandle` and zeroing out) into a method.

The point being `tp_clear` has certain semantics, so it is easier to understand code containing a call to it, rather than has same actions inline.

Also, if `RefCount` will correctly return 0, I'd move this `if` below.

Reply:
Not exactly, `tp_clear` is used to clear the Python object reference in order to break reference cycles. These codes are used to disconnect the relation between Python and clr, if the RefCount reached 0, their relation would naturally be gone.
Their relation needs to be kept if  RefCount > 1 because in `tp_dealloc`, it needs to obtain the bound object.
`tp_dealloc` didn't zero out the handle because this chunk of memory would be freed soon after `tp_dealloc`, extract these into a method is not really necessary.